fix(core): cannot parse ARNs with `/` in resource name

rix0rrr opened the pull request:
Some resource ARNs have "path"-like names, with `/` in them.

We used to do simple splits on separators, which made it impossible
to properly parse entire paths. Example:

```js
'arn:aws:iam::111111111111:role/path/to/role/name'

// split on ':', take 6th element
'role/path/to/role/name'

// split on '/', take 1st or 2nd element
['role', 'path']
```

You see we can get `'role'` and we can get `'path'`, but we cannot get
`'path/to/role/name'` in this way.

However, once we have `'role'`, something else we can do is split on
`':role/'` and take the 2nd element of the resulting array. This will be
the entire path after all!

```js
'arn:aws:iam::111111111111:role/path/to/role/name'

// split on ':role/'
['arn:aws:iam::111111111111', 'path/to/role/name']
                               ~~~~~~~~~~~~~~~~~~  get this
```

We rely on the fact that the `:` anchors the split string to the
start of the path component; it's unlikely that a `:` is going to
occur anywhere else in a resource name.

The resulting expression is fiercely complex. Don't stare at it too
much or it might fry your brain. It does calculate the right value,
which is the most important thing:

```js
{
  'Fn::Select': [1, {
    'Fn::Split': [
      {
        'Fn::Join': ['', [
          ':',
          { 'Fn::Select': [0, { 'Fn::Split': ['/', { 'Fn::Select': [5, { 'Fn::Split': [':', theToken] }] }] }] },
          '/',
        ]],
      },
      theToken,
    ],
  }],
}
```

Fixes #11705


----

*By submitting this pull request, I confirm that my contribution is made under the terms of the Apache-2.0 license*

njlynch reviewed (commented):
Creative!

The changes themselves look fine. However, I have to ask -- should we treat this as a backwards-incompatible change and release under a feature flag? It seems like we're going to change the behavior of several IAM import methods (for the better); however, if anyone is relying on the broken behavior (e.g., manually appending lost paths to group or role names), this could break them.

Also, is there any possibility of resource replacement based on this? If a role/group name changes in the underlying template, might that cause some upstream resource to be recreated/replaced?

(Just being paranoid all around.)

aws-cdk-automation commented:
### AWS CodeBuild CI Report

* CodeBuild project: AutoBuildProject6AEA49D1-qxepHUsryhcu
* Commit ID: 21ec90d49d7c4fab8f9c3e97b0778fca642231bc
* Result: SUCCEEDED
* [Build Logs](https://6sc75qyfl7.execute-api.us-east-1.amazonaws.com/Prod/buildlogs?key=5add9f5c-fb65-4164-a993-3e3ffa16b771%2Fbuild.log) (available for 30 days)

*Powered by [github-codebuild-logs](https://serverlessrepo.aws.amazon.com/applications/arn:aws:serverlessrepo:us-east-1:277187709615:applications~github-codebuild-logs), available on the [AWS Serverless Application Repository](https://aws.amazon.com/serverless/serverlessrepo/)*


<!--
CREATED BY GITHUB-CODEBUILD-LOGS
-->

iliapolo commented:
@njlynch 

> however, if anyone is relying on the broken behavior (e.g., manually appending lost paths to group or role names), this could break them.

👍

rix0rrr commented:
> (Just being paranoid all around.)

Yup, probably pays to be paranoid.

> if anyone is relying on the broken behavior

Doing my best, but I'm having trouble seeing how this would happen though.

> manually appending lost paths to group or role names

It's going to be tough to even do that. Most of our L2s accept an `IRole`; in order to even manually append some fragments to the ARN, you would have had to make your own implementation of `IRole`. 

Take into account that the new behavior will only trigger if the FULL string is a token (not if it's a mostly-literal string with some tokens interspersed). This will realistically only happen if the ARN comes from a `CfnParameter` or something. If so, you would be hard-pressed to know what to suffix the initial part of the path with. And if you DID know the "correct" ARN, you would just funnel the whole string in somewhere rather than relying on a broken `Role.fromRoleArn()` and add in some of your own string manipulation.

For importing roles, I think our current stance is "paths are not supported" (again: for full-token ARNs, not for mostly string-literal ARNs). I have no qualms about breaking an already broken/unsupported scenario. I think it's unlikely that people will have worked around it at all.

> Also, is there any possibility of resource replacement based on this? If a role/group name changes in the underlying template, might that cause some upstream resource to be recreated/replaced?

Extremely unlikely.

* For roles without a path, this will produce the same value, so no replacement.
* For roles WITH a path, the previous code would not have worked, so there's nothing to be replaced either. And if people *did* work around it by manually suffixing the first part of the path (again: unlikely), the new template wouldn't deploy because the role name would now be incorrect, and so no replacement would actually occur.

rix0rrr commented:
Good news everyone!

![image](https://user-images.githubusercontent.com/524162/100618807-982b2f80-331c-11eb-898b-a31871069324.png)

Of course it doesn't actually work.

```
Template error: every Fn::Split object requires two parameters, (1) a string delimiter and (2) a string to be split or a function that returns a string to be split.
```

> For the Fn::Split delimiter, you cannot use any functions. You must specify a string value.
>
> https://docs.aws.amazon.com/AWSCloudFormation/latest/UserGuide/intrinsic-function-reference-split.html

iliapolo commented:
Oooooof course...

rix0rrr commented:
Well that was a waste of half a day.

rix0rrr commented:
Superseded by https://github.com/aws/aws-cdk/pull/11789
